libs/parallelism/execution/include/hpx/execution/algorithms/ensure_started.hpp

```diff
@@ -0,0 +1,379 @@
+                    util::allocator_deleter<other_allocator>>;
+
+                other_allocator alloc(a);
+                unique_ptr p(allocator_traits::allocate(alloc, 1),
+                    hpx::util::allocator_deleter<other_allocator>{alloc});
+
+                new (p.get()) shared_state{std::forward<S_>(s), a};
+                st = p.release();
+
+                // Eagerly start the work received until this point.
+                //
+                // P1897r3 says "When start is called on os2 [the operation
+                // state resulting from connecting the ensure_started_sender to
+                // a receiver], call execution::start(os [the operation state
+                // resulting from connecting S to ensure_started_receiver])",
+                // which would indicate that start should be called later.
+                // However, to fulfill the promise that ensure_started actually
+                // eagerly submits the sender S we call start already here.
+                st->start();
+            }
+
+            ensure_started_sender(ensure_started_sender const&) = default;
+            ensure_started_sender& operator=(
+                ensure_started_sender const&) = default;
+            ensure_started_sender(ensure_started_sender&&) = default;
+            ensure_started_sender& operator=(ensure_started_sender&&) = default;
+
+            template <typename R>
+            struct operation_state
+            {
+                std::decay_t<R> r;
+                hpx::intrusive_ptr<shared_state> st;
+
+                template <typename R_>
+                operation_state(R_&& r, hpx::intrusive_ptr<shared_state> st)
+                  : r(std::forward<R_>(r))
+                  , st(std::move(st))
+                {
+                }
+
+                void start() noexcept
+                {
+                    st->add_continuation(std::move(r));
+                }
+            };
+
+            template <typename R>
+            operation_state<R> connect(R&& r) &&
+            {
+                return {std::forward<R>(r), std::move(st)};
+            }
+
+            template <typename R>
+            operation_state<R> connect(R&& r) &
+            {
+                return {std::forward<R>(r), st};
+            }
+        };
+    }    // namespace detail
+
+    HPX_INLINE_CONSTEXPR_VARIABLE struct ensure_started_t final
+      : hpx::functional::tag_fallback<ensure_started_t>
+    {
+    private:
+        template <typename S,
+            typename Allocator = hpx::util::internal_allocator<>>
+        friend constexpr HPX_FORCEINLINE auto tag_fallback_invoke(
+            ensure_started_t, S&& s, Allocator const& a = {})
+        {
+            return detail::ensure_started_sender<S, Allocator>{
+                std::forward<S>(s), a};
+        }
+
+        template <typename S, typename Allocator,
+            typename NewAllocator = hpx::util::internal_allocator<>>
```

Comment:
What is the `NewAllocator` being used for?

Reply:
That's a very good question. The idea behind this overload is that `ensure_started(ensure_started(sender))` would be the same as `ensure_started(sender)`. `NewAllocator` was meant to be the allocator type for the second (outer) `ensure_started` but be ignored. However, I've used the wrong template parameter in the overload.

In general, I'll admit I'm not entirely sure if this is the right decision (to ignore the new allocator), but I'd find it weird to require them to be the same as well. Do you have a preference or prior art for something like this?

Reply:
I'm not sure I fully understand the implications. It just struck me that there is a (defaulted) template argument that is not used anywhere.